File: src/nfc_reader.py
import nfc
import nfc.clf

class NFCReader:
    """
    A class to handle NFC communication via USB cable using a telephone.
    """

    def __init__(self):
        self.clf = None

    def connect(self):
        """
        Connect to the NFC reader.
        """
        try:
            self.clf = nfc.ContactlessFrontend('usb')
        except Exception as e:
            print(f"Error connecting to NFC reader: {e}")
            self.clf = None

    def read(self):
        """
        Read data from the NFC reader.
        """
        if self.clf is None:
            print("NFC reader is not connected.")
            return None

        try:
            tag = self.clf.connect(rdwr={'on-connect': lambda tag: False})
            return tag
        except Exception as e:
            print(f"Error reading from NFC reader: {e}")
            return None

    def disconnect(self):
        """
        Disconnect from the NFC reader.
        """
        if self.clf is not None:
            self.clf.close()
            self.clf = None

File: src/main/scala/NFCReader.scala
import org.jmrtd.PassportService
import org.jmrtd.lds.icao.MRZInfo
import javax.smartcardio._

class NFCReader {
  private var cardTerminal: CardTerminal = _
  private var card: Card = _
  private var cardChannel: CardChannel = _

  def connect(): Unit = {
    val terminalFactory = TerminalFactory.getDefault
    val terminals = terminalFactory.terminals.list
    if (terminals.isEmpty) {
      throw new Exception("No NFC terminals found")
    }
    cardTerminal = terminals.get(0)
    card = cardTerminal.connect("*")
    cardChannel = card.getBasicChannel
  }

  def connectToAndroidPhone(): Unit = {
    val terminalFactory = TerminalFactory.getDefault
    val terminals = terminalFactory.terminals.list
    if (terminals.isEmpty) {
      throw new Exception("No NFC terminals found")
    }
    cardTerminal = terminals.find(_.getName.contains("Android")).getOrElse(throw new Exception("No Android phone found"))
    card = cardTerminal.connect("*")
    cardChannel = card.getBasicChannel
    println("Connected to Android phone")
  }

  def read(): Array[Byte] = {
    val command = Array[Byte](0x00, 0xA4, 0x04, 0x0C, 0x07, 0xA0, 0x00, 0x00, 0x02, 0x47, 0x10, 0x01)
    val response = cardChannel.transmit(new CommandAPDU(command))
    if (response.getSW != 0x9000) {
      throw new Exception("Failed to read NFC tag")
    }
    response.getData
  }

  def readFromAndroidPhone(): Array[Byte] = {
    val command = Array[Byte](0x00, 0xA4, 0x04, 0x0C, 0x07, 0xA0, 0x00, 0x00, 0x02, 0x47, 0x10, 0x01)
    val response = cardChannel.transmit(new CommandAPDU(command))
    if (response.getSW != 0x9000) {
      throw new Exception("Failed to read NFC tag from Android phone")
    }
    response.getData
  }

  def disconnect(): Unit = {
    if (card != null) {
      card.disconnect(false)
    }
  }
}
